### ISSUE TITLE
Replace AST classifier with model suited for isolated instrument stems

### ISSUE THREAD

**Issue by vsandvold** [vsandvold/mawimbi#249]
## Problem

The instrument classifier assigns the same label to every track. The current model (`Xenova/ast-finetuned-audioset-10-10-0.4593`) is an AudioSet general audio event classifier — trained on 527 broad categories like "Music" and "Musical instrument", not on distinguishing instruments within isolated stems. When fed stems, it returns generic labels that fall through the `instrumentLabels.ts` mapping to the `'percussion'` default.

## Recommended alternative: CLAP zero-shot classification

[CLAP](https://github.com/LAION-AI/CLAP) (Contrastive Language-Audio Pretraining) matches audio embeddings against free-text label embeddings. Instead of a fixed AudioSet label set, you provide candidate labels that match the app's categories exactly — eliminating the `instrumentLabels.ts` mapping layer entirely.

**Models (ONNX, Transformers.js–ready):**

| Model | Audio | Text | Combined (quantized) | Notes |
|---|---|---|---|---|
| [`Xenova/clap-htsat-unfused`](https://huggingface.co/Xenova/clap-htsat-unfused) | 34 MB | 127 MB | **161 MB** | General audio+music |
| [`Xenova/larger_clap_music_and_speech`](https://huggingface.co/Xenova/larger_clap_music_and_speech) | 78 MB | 127 MB | **205 MB** | Music-trained variant |

For reference, the current AST model is **60 MB** (q4).

**Usage:**
```js
const classifier = await pipeline(
  'zero-shot-audio-classification',
  'Xenova/clap-htsat-unfused',
  { device: 'wasm', dtype: 'q8' }
);
const result = await classifier(audio, [
  'bass guitar', 'electric guitar', 'drum kit', 'singing vocals',
  'synthesizer', 'acoustic piano', 'orchestral strings',
  'brass instrument', 'woodwind instrument', 'percussion'
]);
```

## Model size and UX considerations

CLAP is 2–3x larger than the current AST model. This matters for first-load experience since the model must be downloaded before classification can run. Mitigation strategies:

1. **Lazy loading** — the current architecture already loads the model in a Web Worker on first classification, not at app startup. No change needed here.
2. **Stale-while-revalidate caching** — the current `ModelCache` + Cache API strategy already handles this. After the first download, subsequent sessions load from cache instantly and revalidate in the background. The size penalty only affects the *first* visit.
3. **Progressive UX** — show a download progress indicator on first classification instead of just "Detecting instrument…". Transformers.js `pipeline()` accepts a `progress_callback` that reports download bytes. This turns a ~10s wait into an expected, visible process.
4. **Start with the smaller model** — use `clap-htsat-unfused` (161 MB quantized) first. Only consider the larger music-trained variant if accuracy is insufficient.
5. **Prefer quantized text encoder** — CLAP's text encoder (127 MB quantized) dominates the download. Since the candidate labels are fixed strings, the text embeddings could potentially be precomputed and shipped as a small static file (~5 KB for 10 labels × 512 dims), avoiding the text encoder download entirely. This would reduce the download to just the audio encoder (~34 MB quantized — smaller than the current AST model).

## Other alternatives considered

| Option | Download | Pros | Cons |
|---|---|---|---|
| **[onnx-community/Musical-Instrument-Classification-ONNX](https://huggingface.co/onnx-community/Musical-Instrument-Classification-ONNX)** | ~95 MB | 93% accuracy, 9 instrument classes, drop-in pipeline swap | Missing vocals and synth classes; tiny training set (200 samples/class) |
| **Essentia.js + MusiCNN** | ~20 MB model | 40 instrument classes, WASM backend | Adds WASM + TF.js dependencies; less actively maintained |
| **Fine-tune AST on NSynth/IRMAS** | ~60 MB (q4) | Full control, same infrastructure | Requires GPU training + ONNX export; ongoing maintenance |
| **Meyda.js spectral heuristics** | ~20 KB | No model download; instant | Low accuracy; brittle rules |

## Suggested implementation steps

1. Swap `pipeline('audio-classification', ...)` for `pipeline('zero-shot-audio-classification', 'Xenova/clap-htsat-unfused')` in worker and fallback path
2. Pass candidate labels as the second argument instead of `{ top_k: 1 }`
3. Remove `instrumentLabels.ts` mapping — CLAP returns the candidate label directly
4. Add `progress_callback` to surface model download progress in the UI
5. Evaluate accuracy on representative stems; if insufficient, try `larger_clap_music_and_speech`
6. (Optional) Precompute text embeddings for the fixed label set to skip the text encoder download

**Comment by vsandvold** [comment]
Implemented the CLAP zero-shot classification recommendation in PR #250.

### What was done

1. **Swapped model and pipeline** — Replaced `Xenova/ast-finetuned-audioset-10-10-0.4593` (`audio-classification`) with `Xenova/clap-htsat-unfused` (`zero-shot-audio-classification`) in both the worker (`classification.worker.ts`) and main-thread fallback (`InstrumentClassificationService.ts`). Using `q8` quantization as recommended for CLAP.

2. **Replaced label mapping with candidate labels** — Simplified `instrumentLabels.ts` from 60+ AudioSet mappings to 10 descriptive candidate labels (`'singing vocals'`, `'electric guitar'`, `'bass guitar'`, etc.) that map 1:1 to the app's `InstrumentLabel` categories. Exported `CANDIDATE_LABELS` array for use by both worker and service.

3. **Added download progress tracking** — Added `progress_callback` to the worker's `pipeline()` call during network downloads. Worker sends `download-progress` messages to the main thread. `InstrumentClassificationService` exposes a `downloadProgress` signal. Channel UI shows the percentage in the instrument area during first-time model download instead of a generic spinner.

4. **Updated sample rate** — Changed preprocessing target from 16kHz (AST) to 48kHz (CLAP) in both worker and main-thread paths.

5. **Updated all tests** — Replaced AudioSet labels with CLAP candidate labels, added download progress tests (3 new tests), updated assertions for the new pipeline configuration. All 775 tests pass.

### Recommended next steps

1. **Manual accuracy evaluation** — Test classification accuracy on representative instrument stems (isolated vocals, guitar, bass, drums, etc.) and compare against the old AST model
2. **Try larger model if needed** — If accuracy is insufficient, swap to `Xenova/larger_clap_music_and_speech` (205 MB) by changing `MODEL_ID` in the worker and service
3. **Precompute text embeddings** — The CLAP text encoder (127 MB) dominates the download. Since candidate labels are fixed, precomputing text embeddings as a static file (~5 KB) would reduce the download to just the audio encoder (~34 MB), smaller than the old AST model

**Comment by vsandvold** [comment]
### What was done

The CLAP model (`Xenova/clap-htsat-unfused`) was biased toward classifying every uploaded stem as "singing vocals". Instead of swapping to another ML model, added **filename-based classification as the primary detection method** — instant, no model download, and accurate for the majority of real-world use cases where stems have descriptive names.

**Files modified:**
- `src/services/instrumentLabels.ts` — Added `classifyFromFilename()` with keyword patterns for all 10 instrument categories. Normalizes separators and is case-insensitive.
- `src/services/InstrumentClassificationService.ts` — `classify()` now accepts optional `fileName`; tries filename match first (score 1.0), falls back to CLAP only when no keyword matches.
- `src/services/TrackService.ts` — `createTrack()` accepts optional `fileName` and threads it through the `onTrackCreated` callback.
- `src/services/AudioService.ts` — Passes `fileName` from `onTrackCreated` to `classify()`.
- `src/hooks/useTrackService.ts` — Updated `createTrack` bridge to accept optional filename.
- `src/components/project/projectPageEffects.ts` — Passes `file.name` to `createTrack()`.

46 new tests added, all 831 tests pass. PR: #259

### Recommended next steps

1. **Manual accuracy testing** — Upload stems with both descriptive filenames (`bass_guitar.wav`) and generic filenames (`track_01.wav`) to verify filename classification and CLAP fallback paths
2. **Evaluate CLAP fallback quality** — For files with generic names, CLAP still runs. If it still consistently misclassifies, consider removing CLAP entirely and showing "unknown" for unrecognized filenames
3. **Precompute CLAP text embeddings** — If keeping CLAP as fallback, precomputing text embeddings for the fixed label set would reduce the download from 161 MB to ~34 MB

**Comment by vsandvold** [comment]
### What was done

The CLAP model was biased toward classifying every stem as "singing vocals" because its audio encoder processes a fixed ~7-second window, truncating from the start. Stems often begin with silence or quiet intros, so the model received non-representative audio.

**Fix:** Added `extractLoudestSegment()` — a sliding-window function that finds the 10-second segment with the highest RMS energy in the audio. This is applied in both the worker and main-thread fallback classification paths before audio reaches the CLAP model. The approach works for both uploaded files and recorded audio, unlike the filename-based approach in PR #259.

**Files added/modified:**
- `src/services/audioSegment.ts` — New module with `extractLoudestSegment()` (1-second hop sliding window over mono samples)
- `src/services/classification.worker.ts` — Applies segment extraction after downmix/resample, before inference
- `src/services/InstrumentClassificationService.ts` — Applies segment extraction in main-thread fallback path
- `src/services/__tests__/audioSegment.test.ts` — 8 tests covering short passthrough, loudest region selection, edge cases
- `src/services/__tests__/InstrumentClassificationService.test.ts` — 2 new tests for long/short audio in fallback path

PR: #261

### Recommended next steps

1. **Manual accuracy evaluation** — Test with representative stems (isolated vocals, guitar, bass, drums) to verify classification accuracy has improved with the segment extraction
2. **Try the music-trained variant** — If accuracy is still insufficient, swap to `Xenova/larger_clap_music_and_speech` (205 MB) by changing `MODEL_ID` in the worker and service
3. **Precompute text embeddings** — The CLAP text encoder (127 MB) dominates the download. Since candidate labels are fixed, precomputing text embeddings as a static file (~5 KB) would reduce the download to just the audio encoder (~34 MB)

**Comment by vsandvold** [comment]
### What was done

Switched from the general-purpose CLAP model (`Xenova/clap-htsat-unfused`) to the music-trained variant (`Xenova/larger_clap_music_and_speech`) as suggested in the issue. The general-purpose model was biased toward classifying every stem as "singing vocals" because it wasn't trained specifically on music content.

**Files modified:**
- `src/services/classification.worker.ts` — Updated `MODEL_ID` constant to `Xenova/larger_clap_music_and_speech`
- `src/services/InstrumentClassificationService.ts` — Updated model ID in main-thread fallback `pipeline()` call
- `src/services/__tests__/InstrumentClassificationService.test.ts` — Updated test to verify the music-trained model is loaded

All 766 tests pass. PR: #266

### Recommended next steps

1. **Manual accuracy evaluation** — Test classification with representative isolated stems (vocals, guitar, bass, drums, synth) to verify the music-trained model improves accuracy
2. **Precompute text embeddings** — The CLAP text encoder (127 MB) dominates the download. Since candidate labels are fixed, precomputing text embeddings as a static file (~5 KB) would reduce the download from 205 MB to just the audio encoder (~78 MB)
3. **Add download progress indicator** — The music-trained model is larger (205 MB vs 161 MB); consider improving the first-load UX with more detailed progress feedback

**Comment by vsandvold** [comment]
### Option 3: Essentia.js + WASM with Discogs-EffNet + MTG-Jamendo instrument classification

The CLAP model (both general and music-trained variants) has a systemic bias toward "singing vocals" for isolated stems — this persisted through four iterations (PRs #250, #259, #261, #266). CLAP's embedding space was trained on audio-text pairs for general audio, not on discriminating between isolated instrument stems.

**Proposed alternative: Essentia.js WASM feature extraction + ONNX-based instrument classification.**

**Pipeline:**
1. Audio → mono → resample to 16 kHz
2. Essentia.js WASM computes mel spectrogram (96 mel bands, frame=512, hop=256, log compression)
3. `discogs-effnet-bsdynamic-1.onnx` (~18 MB) produces 1280-dim embeddings from mel spectrogram patches
4. `mtg_jamendo_instrument-discogs-effnet-1.onnx` (~2.7 MB) classifies embeddings into 40 instrument categories
5. Map 40 Jamendo classes → app's 10 InstrumentLabel categories

**Why this should work better than CLAP:**
- **Purpose-built for instruments**: The MTG-Jamendo instrument model was trained specifically on instrument classification (40 classes, 0.78 ROC-AUC), not general audio-text matching
- **No zero-shot bias**: Fixed label set from supervised training — no text embedding similarity issues
- **Comprehensive classes**: Covers vocals (`voice`), guitar variants, bass, drums, keyboard, strings, brass, woodwind, synth, and percussion
- **Reasonable download**: ~20.7 MB total (vs 205 MB for CLAP music-trained), model already has ONNX variants

**40 Jamendo instrument classes → 10 app categories:**
- `vocals` ← voice
- `guitar` ← acousticguitar, classicalguitar, electricguitar, guitar
- `bass` ← acousticbassguitar, bass, doublebass
- `drums` ← drums, drummachine
- `keyboard` ← accordion, electricpiano, keyboard, organ, piano, pipeorgan, rhodes
- `strings` ← cello, harp, orchestra, strings, viola, violin
- `brass` ← brass, horn, trombone, trumpet
- `woodwind` ← clarinet, flute, harmonica, oboe, saxophone
- `synth` ← computer, pad, sampler, synthesizer
- `percussion` ← beat, bell, bongo, percussion

**Dependencies added:**
- `essentia.js` (~2.7 MB packed) — WASM backend for mel spectrogram computation
- `onnxruntime-web` (already a transitive dep of `@huggingface/transformers`) — ONNX model inference

**Trade-offs vs other alternatives:**
- Filename-based classification (Option 1) is faster/simpler but doesn't work for recorded audio or generic filenames
- Spectral heuristics (Option 2) need no model download but have lower accuracy
- This approach (Option 3) gives ML-grade accuracy with a purpose-built model at reasonable download size
